### PR TITLE
[FLINK-36535][autoscaler] Optimize the scale down logic based on historical parallelism to reduce the rescale frequency

### PR DESCRIPTION
**Pull request by 1996fanrui**
## What is the purpose of the change

Get more details from FLINK-36535.

## Brief change log

[FLINK-36535][autoscaler] Optimize the scale down logic based on historical parallelism
1. Using the maximum parallelism within the window instead of the latest parallelism when scaling down
2. Never scale down when (currentTime - triggerTime) < scale-down.interval

## Verifying this change

This change added tests and can be verified as follows:

- Added the DelayedScaleDownTest
- Rewrite the `testRequiredScaleDownAfterInterval` to `testScaleDownAfterInterval`
- Update the `AbstractAutoScalerStateStoreTest`, `JobVertexScalerTest` and `ScalingExecutorTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Review by gyfora — COMMENTED** [review]
I think one other problem with this logic in general is that we are handling the scale down window/interval more or less independently for different vertices.

So if you have 10 vertices and they would be scaled down at different times you can have 10 restarts within the scale down window. Which does not feel right....

I think we should consider setting global scale down windows, for example one possible implementation would be that if the scale down interval is 1 hour and we have one vertex requesting the scale down first at 12:10 the second at 12:20, then we scale both down at 13:20 (instead of doing 2 scale downs). So basically coalescing the intervals to some extent to really max 1 scale down per hour

**Comment by mxm** [comment]
> So basically coalescing the intervals to some extent to really max 1 scale down per hour

Exactly, I actually thought that's how the originally proposed "lazy" scale down technique should work.

**Comment by 1996fanrui** [comment]
Thanks @gyfora and @mxm for the quick comment and suggestion!

> So if you have 10 vertices and they would be scaled down at different times you can have 10 restarts within the scale down window. Which does not feel right.

Actually, we only rescale twice (instead of 10 times) if we have 10 vertices and they would be scaled down at different times.
Assuming scale down interval is 1 hour:

- vertex1 triggers scale down at 12:10,
- vertex2 triggers scale down at 12:15,
- vertex3 triggers scale down at 12:18,
- vertex10 triggers scale down at 12:50,

And then vertex1 will be scaled down at 13:10, and rest of them do not be changed.

After 13: 10, if all of vertex2 to vertex10 need to scale down,  all of them will trigger scale down at 13: 10, and will be scaled down at 14:10

> I think we should consider setting global scale down windows, for example one possible implementation would be that if the scale down interval is 1 hour and we have one vertex requesting the scale down first at 12:10 the second at 12:20, then we scale both down at 13:20 (instead of doing 2 scale downs). So basically coalescing the intervals to some extent to really max 1 scale down per hour

There may be an unexpected case, and it’s easy to happen when scale down interval >= 24 hours.
We assume scale down interval = 24 hours, and the peak hour are following:

- vertex1 18:00
- vertex2 20:00
- vertex3 22:00

Trigger times:

- vertex1 triggers scale down at 2024-11-20 19:00,
- vertex2 triggers scale down at 2024-11-20 21:00,
- vertex3 triggers scale down at 2024-11-20 23:00,

So this strategy hope scale down is executed at 2024-11-21 23:00. If all of vertices need to scale down after 24 hours, it works well.

But unexpected case is : vertex2 always wanna scale down, but vertex1 and vertex3 runs on the parallelism of peak time.

The scale down trigger is canceled for vertex1 at 2024-11-21 18:00,  and
vertex1 re-triggered scale down at 2024-11-21 19:00.

At this time, the trigger status are:

- vertex1 triggers scale down at 2024-11-21 19:00,
- vertex2 triggers scale down at 2024-11-20 21:00,
- vertex3 triggers scale down at 2024-11-20 23:00,

It hopes scale down is executed at 2024-11-22 19:00

But vertex3 will cancel and re-trigger again, the execution time will be postponed again.
It will be looped everyday, and scale down of vertex2 never happen.

Please correct me if I misunderstand anything, and I’m happy to hear more suggestions from you.

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/DelayedScaleDown.java [hunk 1]
```diff
@@ -19,51 +19,81 @@
...
     void clearAll() {
-        if (firstTriggerTime.isEmpty()) {
+        if (delayedVertices.isEmpty()) {
             return;
         }
-        firstTriggerTime.clear();
-        isUpdated = true;
+        delayedVertices.clear();
+        updated = true;
     }
```
Comment (1996fanrui) [inline flink-autoscaler/src/main/java/org/apache/flink/autoscaler/DelayedScaleDown.java:98]: We clear all delayed scale down when rescale happens, so the scale down won't happen for vertices (that trigger time is less than scale down interval).

Ideally, these vertices will re-trigger scale down after rescaling. It likes this demo:

> So if you have 10 vertices and they would be scaled down at different times you can have 10 restarts within the scale down window. Which does not feel right.

Actually, we only rescale twice (instead of 10 times) if we have 10 vertices and they would be scaled down at different times.
Assuming scale down interval is 1 hour:

- vertex1 triggers scale down at 12:10,
- vertex2 triggers scale down at 12:15,
- vertex3 triggers scale down at 12:18,
- vertex10 triggers scale down at 12:50,

And then vertex1 will be scaled down at 13:10, and rest of them do not be changed.

After 13: 10, if all of vertex2 to vertex10 need to scale down, all of them will trigger scale down at 13: 11 (next reconcile loop), and will be scaled down at 14:11 .

**Review by 1996fanrui — COMMENTED** [review]
I added some comments to help understand this approach easily.

**Comment by gyfora** [comment]
Thanks @1996fanrui , after some offline discussion I think this works as expected I just slightly misunderstood the mechanism. 

We can later optimize this further by coalescing downscale triggers that are close to each other but we don't need to do it right now.

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/ScalingExecutor.java [hunk 2]
```diff
@@ -181,15 +178,15 @@ private void updateRecommendedParallelism(
...
-        // All vertices' ParallelismChange is optional, rescaling will be ignored.
-        if (requiredVertices.isEmpty()) {
+            Set<JobVertexID> changedVertices) {
+        // No any vertex is changed.
+        if (changedVertices.isEmpty()) {
             return true;
         }
 
-        for (JobVertexID vertex : requiredVertices) {
+        for (var vertex : changedVertices) {
```
Comment (mxm) [inline flink-autoscaler/src/main/java/org/apache/flink/autoscaler/ScalingExecutor.java:189]: NIT Why this change? Makes the loop harder to read.

**Review by mxm — APPROVED** [review]
LGTM

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java [hunk 3]
```diff
@@ -289,21 +274,26 @@ private ParallelismChange applyScaleDownInterval(
...
+                LOG.debug(
+                        "Try to skip immediate scale down within scale-down interval for {}",
+                        vertex);
+            }
+            return ParallelismChange.noChange();
         } else {
-            return ParallelismChange.required(newParallelism);
+            // Using the maximum parallelism within the scale down interval window instead of the
+            // latest parallelism when scaling down
+            return ParallelismChange.build(delayedScaleDownInfo.getMaxRecommendedParallelism());
```
[the remark this answers is not in the document]
Comment (mxm) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java:296]: Can we make sure though, that the recommended scale down parallelism is not higher than the current one? This could theoretically happen if the recommended parallelism is higher than the current parallelism during the scale down period.

**Review by huyuanfeng2018 — COMMENTED** [review]
a little bit comment, PTAL

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java [hunk 4]
```diff
@@ -88,31 +88,22 @@ public JobVertexScaler(AutoScalerEventHandler<KEY, Context> autoScalerEventHandl
...
-
-    /**
-     * The rescaling will be triggered if any vertex's ParallelismChange is required. This means
-     * that if all vertices' ParallelismChange is optional, rescaling will be ignored.
-     */
+    /** The rescaling will be triggered if any vertex's {@link ParallelismChange} is changed. */
     @Getter
     public static class ParallelismChange {
 
-        private static final ParallelismChange NO_CHANGE =
```
[the remark this answers is not in the document]
Comment (1996fanrui) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java:105]: We could keep it in this PR first, and refactor it in the future if needed.

**Review by 1996fanrui — COMMENTED** [review]
Thanks @gyfora for the update!

And thanks @mxm and @huyuanfeng2018 for the quick review!

I have addressed or replied all comments.

**Review by huyuanfeng2018 — APPROVED** [review]

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java [hunk 5]
```diff
@@ -289,21 +274,26 @@ private ParallelismChange applyScaleDownInterval(
...
         }
 
-        if (clock.instant().isBefore(firstTriggerTime.get().plus(scaleDownInterval))) {
-            LOG.debug("Try to skip immediate scale down within scale-down interval for {}", vertex);
-            return ParallelismChange.optional(newParallelism);
+        var now = clock.instant();
+        var delayedScaleDownInfo = delayedScaleDown.triggerScaleDown(vertex, now, newParallelism);
+
+        // Never scale down within scale down interval
+        if (now.isBefore(delayedScaleDownInfo.getFirstTriggerTime().plus(scaleDownInterval))) {
```
[the remark this answers is not in the document]
Comment (mxm) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java:284]: I thought that `getFirstTriggerTime()` returns the first time we scaled *up*, but we are actually recording the time we first try to scale down. 

I'm not sure this is correct. We want to delay scale down from the first time we scale up, not the first time we scaled down.

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mxm) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java:296]: Thanks Rui!

**Review by mxm — COMMENTED** [review]
Generally, the PR looks good. 

I missed this in the original JIRA, even though you wrote it there: https://github.com/apache/flink-kubernetes-operator/pull/920#discussion_r1861985148 I can open a separate JIRA issue, if we don't want to resolve it here.

**Review by gyfora — APPROVED** [review]

**Review by mxm — APPROVED** [review]
LGTM. We can handle https://github.com/apache/flink-kubernetes-operator/pull/920#discussion_r1861985148 independently of this PR.

**Comment by gyfora** [comment]
please rebase on prod as there is a conflicting test change

**Comment by 1996fanrui** [comment]
Thanks for the review, I will address all comments next Monday due to I'm on the Flink Forward Asia Conference.

File: flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (1996fanrui) [reply flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java:284]: Thanks @mxm  for the clarification!

Actually, we expect to use the first scale down time instead of scale up time as the first trigger time. 

- job.autoscaler.scale-up.grace-period hopes to use  scale up time
- And job.autoscaler.scale-down.interval hopes to use the first scale down time, it `scale-down.interval` is 1 hour, we expect the scale down can be executed after 1 hour.
    - It's delayed scale down
    - And It could merge multiple scale down requests into one scale down execution.
- If scale-down.interval uses the scale up time, when scale down request comes, job will execute scale down directly if job is scaled up 1 hour ago. (It cannot merge multiple scale down request into one execution)

As I understand, the scale down interval wanna to reduce the scale down frequency.

Please correct me if anything is wrong, thank you~

**Comment by 1996fanrui** [comment]
> please rebase on main as there is a conflicting test change

thanks @gyfora for the reminder, rebased and fixed the conflict.

**Comment by 1996fanrui** [comment]
Thanks all for the patient review!

We could discuss https://github.com/apache/flink-kubernetes-operator/pull/920#discussion_r1858987790  asynchronously, so let me merge this PR first.
